gossip/components/responder.c

```diff
@@ -6,219 +6,278 @@
-                                     request->hash->trits, false)) != RC_OK) {
+    // If a transaction was requested but not found, requests it
+    if (!flex_trits_are_null(hash, FLEX_TRIT_SIZE_243) &&
+        rand_handle_probability() < responder->node->conf.p_propagate_request) {
+      if ((ret = request_transaction(&responder->node->transaction_requester,
+                                     hash, false)) != RC_OK) {
+        log_warning(RESPONDER_LOGGER_ID, "Requesting transaction failed\n");
         return ret;
       }
     }
   }
+
   return RC_OK;
 }
 
-static void *responder_routine(responder_state_t *const state) {
-  transaction_request_t request;
-  iota_transaction_t tx = NULL;
+/**
+ * Continuously looks for a transaction request from a responder queue and
+ * process it
+ *
+ * @param responder The responder state
+ */
+static void *responder_routine(responder_t *const responder) {
+  transaction_request_t *request = NULL;
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, pack);
+  trit_array_t hash = {.trits = NULL,
+                       .num_trits = HASH_LENGTH_TRIT,
+                       .num_bytes = FLEX_TRIT_SIZE_243,
+                       .dynamic = 0};
 
-  if (state == NULL) {
+  if (responder == NULL) {
     return NULL;
   }
 
-  while (state->running) {
-    if (CQ_POP(state->queue, &request) == CQ_SUCCESS) {
-      log_debug(RESPONDER_COMPONENT_LOGGER_ID, "Responding to request\n");
-      if (get_transaction_for_request(state, &request, &tx)) {
-        log_warning(RESPONDER_COMPONENT_LOGGER_ID,
-                    "Getting transaction for request failed\n");
-        continue;
-      }
-      if (reply_to_request(state, &request, tx)) {
-        log_warning(RESPONDER_COMPONENT_LOGGER_ID,
-                    "Replying to request failed\n");
-        continue;
-      }
+  lock_handle_t lock_cond;
+  lock_handle_init(&lock_cond);
+  lock_handle_lock(&lock_cond);
+
+  while (responder->running) {
+    if (responder_size(responder) == 0) {
+      cond_handle_timedwait(&responder->cond, &lock_cond,
+                            RESPONDER_TIMEOUT_SEC);
+    }
+    rw_lock_handle_wrlock(&responder->lock);
+    request = transaction_request_queue_peek(responder->queue);
+    if (request == NULL) {
+      rw_lock_handle_unlock(&responder->lock);
+      continue;
+    }
+    log_debug(RESPONDER_LOGGER_ID, "Responding to request\n");
+    hash_pack_reset(&pack);
+    hash.trits = request->hash;
+    if (get_transaction_for_request(responder, request->neighbor, &hash,
+                                    &pack) != RC_OK) {
+      log_warning(RESPONDER_LOGGER_ID,
+                  "Getting transaction for request failed\n");
+    } else if (respond_to_request(responder, request->neighbor, request->hash,
+                                  &pack) != RC_OK) {
+      log_warning(RESPONDER_LOGGER_ID, "Replying to request failed\n");
     }
+    transaction_request_queue_pop(&responder->queue);
```

Comment:
If we don't have the requested tx in our DB we should add it to our own TX req queue I believe?

Either way: we can pop instead of peek here as we always pop? (and thus reducing the time we hold the lock)

Reply:
We do request missing tx [here](https://github.com/iotaledger/entangled/pull/484/files/ec15c796b1cb3c08d3cefaf4bea6b3abf02b5646#diff-6c871dd3004f1b74db143dffd9d4ffe4R101).

Reply:
pop only delete the first element, it doesn't return it, and peek returns a pointer to the first element. What we can do to reduce the lock holding time is `lock, peek, make a copy of the element, pop, unlock, proceed to function` ?